### PR TITLE
docs(core): improve `inject` function docs

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit updates the `inject` function docs by:
- rephrasing a description to include more usage cases
- adding usage examples
- making a function itself a public API (vs its alias const that was used previously)

Before/after comparison:
inject-docs

## PR Type
What kind of change does this PR introduce?

- [x] Documentation content changes

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by pkozlowski-opensource — APPROVED** [review]
This looks so much better now, thank you @AndrewKushnir !

LGTM

**Review by pkozlowski-opensource — APPROVED** [review]
Reviewed-for: public-api

**Review by dylhunn — APPROVED** [review]

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:452096996:BASE:452097062:1654021462234:1240ee78).

File: packages/core/test/bundling/animations/bundle.golden_symbols.json [hunk 1]
```diff
@@ -941,9 +941,6 @@
   {
     "name": "initTNodeFlags"
   },
-  {
-    "name": "inject"
```
Comment (AndrewKushnir) [inline packages/core/test/bundling/animations/bundle.golden_symbols.json:945]: This change is likely a result of switching:
```
export const inject = ɵɵinject;
```
to:
```
export function inject<T>(token: ProviderToken<T>, flags = InjectFlags.Default): T|null {
  return ɵɵinject(token, flags);
}
```
Not sure why the `inject` got retained previously, I don't think there was any code in the bundle that referenced it.

File: packages/core/src/di/injector_compatibility.ts [hunk 2]
```diff
@@ -109,30 +106,102 @@ Please check that 1) the type for the parameter at index ${
...
+ * ```typescript
+ * @Injectable({providedIn: 'root'})
+ * export class Car {
+ *   radio: Radio|undefined;
+ *   // OK: field initializer
+ *   spareTyre = inject(Tyre);
  *
- * {@example core/di/ts/injector_spec.ts region='ShakableInjectionToken'}
+ *   // OK: constructor parameter
+ *   constructor(readonly engine: Engine = inject(Engine)) {
```
Comment (alxhub) [inline packages/core/src/di/injector_compatibility.ts:150]: We should remove this pattern - `@Injectable` will try to compile a factory function for this constructor, and so the parameter initializer isn't actually doing anything here.

File: packages/core/src/di/injector_compatibility.ts [hunk 3]
```diff
@@ -109,30 +106,102 @@ Please check that 1) the type for the parameter at index ${
...
+ *     return new Car(engine);
+ *   }}
+ * ]
+ * ```
+ *
+ * The `inject()` calls might be wrapped in an intermediate function and will work correctly as long
+ * as a wrapper function is used in one of the allowed contexts:
+ *
+ * ```typescript
+ * function getAndLog<T>(token: ProviderToken<T>): T {
```
Comment (alxhub) [inline packages/core/src/di/injector_compatibility.ts:173]: We shouldn't encourage people to wrap `inject` like this. At the very least, we should change `getAndLog` to be `injectAndLog`.

File: packages/core/src/di/injector_compatibility.ts [hunk 4]
```diff
@@ -109,30 +106,102 @@ Please check that 1) the type for the parameter at index ${
...
+ * }
+ * ```
+ *
+ * Calls to the `inject()` function outside of the class creation context will result in error. Most
+ * notably, calls to `inject()` are disallowed after a class instance was created, in methods:
+ *
+ * ```typescript
+ * @Injectable({providedIn: 'root'})
+ * export class Car {
+ *   start() {
```
Comment (alxhub) [inline packages/core/src/di/injector_compatibility.ts:192]: Maybe change this to `ngOnInit` to highlight that it doesn't work in lifecycle hooks either?

File: packages/core/testing/src/test_bed.ts [hunk 5]
```diff
@@ -107,7 +107,10 @@ export const TestBed: TestBedStatic = TestBedRender3;
 export const getTestBed: () => TestBed = _getTestBedRender3;
 
 /**
- * Allows injecting dependencies in `beforeEach()` and `it()`.
+ * Allows injecting dependencies in `beforeEach()` and `it()`. Note: this function
+ * (imported from the `@angular/core/testing` package) can **only** be used to inject dependencies
+ * in tests. To inject dependencies in your application code, use [the `inject` function from the
```
Comment (alxhub) [inline packages/core/testing/src/test_bed.ts:112]: Link only `inject` here.

File: packages/core/src/di/injector_compatibility.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/core/src/di/injector_compatibility.ts:173]: Ok, I've removed this example.

File: packages/core/src/di/injector_compatibility.ts (hunk 4 above)
Comment (AndrewKushnir) [reply packages/core/src/di/injector_compatibility.ts:192]: 👍

File: packages/core/src/di/injector_compatibility.ts [hunk 6]
```diff
@@ -109,30 +106,102 @@ Please check that 1) the type for the parameter at index ${
...
  *
- * Within such a factory function, using this function to request injection of a dependency
- * is faster and more type-safe than providing an additional array of dependencies
- * (as has been common with `useFactory` providers).
+ * @publicApi
+ */
+export function inject<T>(token: ProviderToken<T>, flags?: InjectFlags): T|null;
+/**
+ * Injects a token from the currently active injector.
+ * The injection context for this function is available during the class creation and
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/core/src/di/injector_compatibility.ts:128]: 👍

**Comment by AndrewKushnir** [comment]
@alxhub thanks for the feedback, I've addressed it in a fixup commit: https://github.com/angular/angular/pull/46168/commits/d87cd8592e39ee62768ca45bc407cd09747dcb05. Could you please take another look when you get a chance? Thank you.

**Review by alxhub — APPROVED** [review]

**Comment by alxhub** [comment]
The code change here is minimal and is intended to ensure that `inject()` is treated correctly by the documentation site (as a `function` and not a `const`). It has no behavioral impact and should be safe to land even during the RC freeze.

**Comment by atscott** [comment]
This PR was merged into the repository by commit 7f65089de865dab893189293d052472ee34309c9.

**Pull request closed without merging**